<div style="margin-left:15%;
            width:70%;
            height:10px;
            background: #edac1a;  
"></div>  
<p style="
           margin-top: 10px;
           margin-bottom: 10px;
             text-align: center;
             font-family: sans-serif;
             font-size: 3rem;
             letter-spacing: 0.15rem;
             text-transform: uppercase;
             color: #2b3e85; 
          font-weight:bold;
"> 
 Hensel's Lemma in Sage
</h1>
<div style="margin-left:15%;
            width:70%;
            height:10px;
            background: #edac1a;  
"></div>

## Hensel's Lemma

> **Hensel's Lemma:** Let $f(x)$ be a polynomial with integer coefficients, $p$ a prime, and $e\geq 1$ an integer.
> Suppose $f(r)\equiv 0\pmod {p^e}$ and $f'(r)\neq 0\pmod p$, then 
>
> $$s=r+kp^{e}$$
>
> is a solution to the congruence $f(x)\equiv 0\pmod{p^{e+1}}$, where $k\in\mathbb{Z}$ satisfies 
>
> $$\frac{f(r)}{p^e}+kf'(r)\equiv 0\pmod p.$$

> 🎯 Complete the proof in the case that $\deg(f)=3$.

**Proof:** First, we prove that $k$ exists.  Note that $\frac{f(r)}{p^e}\in\mathbb{Z}$ because ..................

Further, since $\gcd(p,f'(r))=0$ **why?**, we know that $f'(r)x\equiv b\pmod p$ has exactly 1 solution for all $b\in\mathbb{Z}$.  That is, $k$ exists, as does $s=r+kp^e$.


Assume $\deg(f)=3$. Then we can write $f(x)=a_0+a_1x+a_2x^2+a_3x^3$ for some $a_0,a_1,a_2,a_3\in\mathbb{Z}$. Before computations, we find $f'(r)=...........$.  Now compute

$$\begin{eqnarray*}
f(r+kp^e) &=& a_0+a_1(r+kp^e)+a_2(r+kp^e)^2+a_3(r+kp^e)^3,\\
&=& a_0+a_1(r+kp^e)+a_2(r^2+2rkp^e+k^2p^{2e})+a_3(r^3+3r^2kp^e+3rk^2p^{2e}+k^3p^{3e}),\\
&\equiv& a_0+a_1(r+kp^e)+a_2(r^2+2rkp^e)+a_3(r^3+3r^2kp^e)\pmod{p^{e+1}},\quad \textit{why?}\\
&=& ....................,\\
&=& f(r)+ p^e f'(r),\\
\end{eqnarray*}$$

Since ....................., we know that $f(r)+ p^e f'(r)\equiv 0\pmod{p^{e+1}}$. 
And so we conclude that $f(r+p^e)\equiv 0\pmod{p^{e+1}}$, as desired.
<div style="text-align:right;">🔲</div>

<div style="width:100%;  height:5px;  background: #2b3e85;"></div>

## Hensel's Lemma on polynomials

In class we worked through $3x\equiv 2\pmod{5^e}$ for $e=1,2,3$.  Here's one way I would do that in Sage:

In [1]:
# First roots mod 5:
_.<x> = PolynomialRing(Integers())
f = 3*x-2
f_prime = f.derivative()

r = next(r for r in range(5) if f(r)%5==0)
print("r =",r) # This has only 1 root so we can use the first

assert f(r)%5 == 0

r = 4


In [2]:
# Now we go from 5 to 25:
r = 4 # Taken from above
assert f_prime(r)%5 != 0

k = next(k for k in range(5) if (f(r)/5+f_prime(r)*k)%5==0)
print("k =",k)

s = r + 5*k

print("s =",s)
assert f(s)%25 == 0

k = 1
s = 9


In [3]:
# Now we go from 25 to 125:
r = 9 # Taken from above
assert f_prime()(r)%5 != 0

k = next(k for k in range(5) if (f(r)/25+f_prime(r)*k)%5==0)
print("k =",k)

s = r + 25*k

print("s =",s)
assert f(s)%125 == 0

k = 3
s = 84


In [4]:
# Now we go from 125 to 625:
r = 84 # Taken from above
assert f_prime()(r)%5 != 0

k = next(k for k in range(5) if (f(r)/125+f_prime(r)*k)%5==0)
print("k =",k)
s = r + 125*k

print("s =",s)
assert f(s)%625 == 0

k = 1
s = 209


<div style="width:100%;  height:5px;  background: #2b3e85;"></div>

## Now you do it

> 🎯 Use Sage and Hensel's Lemma to find solutions to $3x^2-5\equiv 0 \pmod {17^e}$ for $e=1,2,3,4$.

<div style="width:100%;  height:5px;  background: #2b3e85;"></div>

## Now you do it, again

> 🎯 Use Sage and Hensel's Lemma to find solutions to $17x^4-32x^3+5x^2-2x+23 \equiv 0 \pmod {11^e}$ for $e=1,2,3,4$.

<div style="width:100%;  height:5px;  background: #2b3e85;"></div>

## Getting funky 🦨

> 🎯 First, explain why Hensel's Lemma does not apply to $f(x)=x^5-1$ for $p=5$ and any $e$.

**Answer:**

> 🎯 Find an $r\in\mathbb{Z}$ that is a root of $f(x)=x^5-1$, and use it to explain why $f(x)\equiv 0\pmod {5^e}$ has a solution for all $e$. 

**Answer:**

<div style="width:100%;  height:5px;  background: #2b3e85;"></div>

## A different funk 🎺

> 🎯 Explain why Hensel's Lemma does not apply to $f(x)=x^2+2x+1$ for $p=3$ and $e=1$.

**Answer:**

> 🎯 Explain why Hensel's Lemma *does* apply to $f(x)=x^2+2x+1$ for $p=3$ and $e=2$ using $r=2$, and use Hensel's Lemma to solve $f(x)\equiv 0\pmod{p^{e+1}}$ given $f(2)\equiv 0\pmod{p^e}$.

**Answer:**

> 🎯 Explain why Hensel's Lemma *does NOT* apply to $f(x)=x^2+2x+1$ for $p=3$ and $e=2$ using $r=8$. Why is this OK?

**Answer:**

<div style="width:100%;  height:5px;  background: #2b3e85;"></div>

## $p$-adic integers

Notice one important thing, $s \equiv r\pmod{p^e}$.

In particular our sequence of solutions $(4,9,84,209)$ are congruenct mod 5, the sequence $(9,84,209)$ are congruent mod 25, and finally $84\equiv 209\pmod{125}$.

This sequence continues forever, $(4,9,84,209,\dots)$, and actually there's a name for sequences like these, $p$-adic integers.  There are many ways to represent these, by default, Sage represents them as sums

$$a_0+a_1\cdot p+a_2\cdot p^2+a_3\cdot p^3+\cdots,$$

where the $a_i$ values are actually the $k$ values that we've been finding.

In our example $(4,9,84,209,\dots)$ would be represented as 

$$4+1\cdot5+3\cdot5^2+1\cdot5^3+\cdots.$$

If we call this number $r$, then 

$$\begin{eqnarray*}
r&\equiv& 4\pmod 5\\
r&\equiv&  9\pmod{25}\\
r&\equiv&  84\pmod{125}\\
r&\equiv&  209\pmod{625}
\end{eqnarray*}$$

Here's how we can construct $p$-adic integers and find roots mod $p^e$.

In [ ]:
R = Zp(5)                         # Define p-adics (choosing p)
_.<x> = PolynomialRing(R)         # Set up x as a polynomial variable
f = 3*x-2                         # Define the actual polynomial
r = f.roots(multiplicities=False)[0]  # Find the roots of the polynomial
print("r =",r)
 
print(f"\n\nSolutions mod 5^e for e from 1 to {r.precision_absolute()-1}")
print([mod(r,5^e) for e in range(1,r.precision_absolute())])

> **Exercise:** 🎯 Last time we also used Hensel's Lemma to solve $x^2+1\equiv 0$ modulo $13$, $13^2$ and $13^3$. Use the code in the previous block to extend this beyond these values, solving $x^2+1\equiv 0\pmod{13^e}$ for $e$ from 1 to 19.

> **Exercise:** 🎯 What does this code produce when we have an integer solution, for example, if $f=x^2-1$ with $p$ your favorite prime?

<div style="width:100%;  height:5px;  background: #2b3e85;"></div>

## 🍰 Optional - Me using Sage to do symbolic computations

In [ ]:
# Choosing the prime, exponent, and degree of f

p = 5
e = 17
degree = 4

# Defining the coefficients of f as a[i] for 0<=i<=degree, they're considered mod p^(e+1)
A = PolynomialRing(Integers(p^(e+1)),degree+1,names="a_")
a = A.gens()

# Defining f as a polynomial with coefficients from A
_.<x> = PolynomialRing(A)

f = sum(a[i]*x^i for i in range(len(a)))

# Defining r and k as variables. We could instead use r=_ and k=_ to define them as actual integers.
_.<r,k> = PolynomialRing(A)

# Comparing polynomials
print("f(r+kp^e)=")
print(f(r+k*p^e))
print("\nf(r)=")
print(f(r))
print("\np^ekf'(r)=")
print(p^e*k*f.derivative()(r))
print('\nAre these equal?')
print(f(r+k*p^e)==(f(r)+p^e*k*f.derivative()(r)))

In [ ]:
# If we want to factor these coefficients we have to work backward 
# through several polynomial rings: (This gets nasty as I try to make it print nice.)
factored_coefficients = [Integers()(n).factor() for c in f(r+k*p^e).coefficients() for n in c.coefficients()]
f_as_string = "" 
monomials = [m for m  in f(r+k*p^e).monomials()]
next_monomials = [m for c in f(r+k*p^e).coefficients() for m in c.monomials()]
def stringify(c):
    if str(c)=="1":
        return ""
    current = "("
    exp = lambda pair: str(pair[0]) if pair[1]==1 else f"{pair[0]}^{pair[1]}"
    current += ")(".join(exp(pair) for pair in c)
    return current+")"
for i in range(len(factored_coefficients)):
    f_as_string += stringify(factored_coefficients[i])
    f_as_string += str(next_monomials[i])
    f_as_string += str(monomials[i])
    f_as_string += " + "


print("f(r+kp^e)=")
print(f_as_string.replace("*","").rstrip(" +"))